### PR TITLE
Modify `derive(Debug)` to use `Self` in struct literal to avoid redundant error

### PR DESCRIPTION
**Pull request by estebank**
Reduce verbosity in #97343.

### CODE REVIEW

File: src/test/ui/derives/issue-97343.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+use std::fmt::Debug;
+
+#[derive(Debug)]
+pub struct Irrelevant<Irrelevant> { //~ ERROR type arguments are not allowed for this type
```
Comment (compiler-errors) [inline src/test/ui/derives/issue-97343.rs:4]: Huh, this error message is very confusing imo. I guess that's not necessarily the purpose of this PR. 
r=me anyways.

**Comment by compiler-errors** [comment]
re the existing error message this doesn't touch, @estebank, should I reword that message to say something like "type argument shadows item name" or something?

File: src/test/ui/derives/issue-97343.rs (hunk 1 above)
Comment (estebank) [reply src/test/ui/derives/issue-97343.rs:4]: Leaving https://github.com/rust-lang/rust/issues/97343 open to both address that (I think a new error code is warranted) and potentially make the macro smarter to make this code work in the first place.

**Comment by estebank** [comment]
@compiler-errors if you can do that, that'd be great. I think that you'll need to do name resolution on the *parent* scope (to avoid the shadowing) and compare the def id of what you find and what we currently have. Once we have that we can go fancy and point at "this type argument..."/"...shadows this item name in the current scope"/"consider renaming the type parameter".

**Pull request merged**